lib/bitcoin/SidetreeTransactionData.ts

```diff
@@ -0,0 +1,131 @@
+
+    let sidetreeDataToReturn: string | undefined = undefined;
+
+    for (let i = 0; i < transactionOutputs.length; i++) {
+      const currentOutput = transactionOutputs[i];
+      const sidetreeDataForThisOutput = SidetreeTransactionData.getSidetreeDataFromOutputIfExist(currentOutput, sidetreePrefix);
+
+      if (sidetreeDataForThisOutput) {
+
+        const oneSidetreeDataAlreadyFound = sidetreeDataToReturn !== undefined;
+
+        if (oneSidetreeDataAlreadyFound) {
+          console.info(`More than one sidetree transactions were found in the outputs of transaction id: ${transactionId}`);
+          return undefined;
+        }
+
+        sidetreeDataToReturn = sidetreeDataForThisOutput;
+      }
+    }
+
+    return sidetreeDataToReturn;
+  }
+
+  private static getSidetreeDataFromOutputIfExist (transactionOutput: BitcoinOutputModel, sidetreePrefix: string): string | undefined {
+
+    // check for returned data for sidetree prefix
+    const hexDataMatches = transactionOutput.scriptAsmAsString.match(/\s*OP_RETURN ([0-9a-fA-F]+)$/);
+
+    if (hexDataMatches && hexDataMatches.length !== 0) {
+
+      const data = Buffer.from(hexDataMatches[1], 'hex').toString();
+
+      if (data.startsWith(sidetreePrefix)) {
+        return data.slice(sidetreePrefix.length);
+      }
+    }
+
+    // Nothing was found
+    return undefined;
+  }
+
+  private static getValidWriterFromInputs (transactionId: string, transactionInputs: BitcoinInputModel[]): string | undefined {
+
+    const validPublickey = SidetreeTransactionData.getValidPublicKeyFromInputs(transactionId, transactionInputs);
+
+    if (!validPublickey) {
+      return undefined;
+    }
+
+    // This library uses the public-key-hash-output format of the public key as the writer so
+    // convert the public key into the expected foramat.
+    const publicKeyAsBuffer = Buffer.from(validPublickey, 'hex');
+    const publicKeyHashBuffer = crypto.Hash.sha256ripemd160(publicKeyAsBuffer);
+
+    return publicKeyHashBuffer.toString('hex');
+  }
+
+  private static getValidPublicKeyFromInputs (transactionId: string, transactionInputs: BitcoinInputModel[]): string | undefined {
+
+    // First get all the public keys from the inputs
+    const allPublicKeys = transactionInputs.map(input => {
+      const scriptAsmParts = input.scriptAsmAsString.split(' ');
+
+      // If the publickey is not present then just use 'undefined'
+      return scriptAsmParts.length >= 2 ? scriptAsmParts[1] : undefined;
+    });
+
+    // Save all the unique public keys.
+    const uniquePublicKeys = new Set<string | undefined>();
+    for (let i = 0; i < transactionInputs.length; i++) {
+      uniquePublicKeys.add(allPublicKeys[i]);
+    }
+
+    // There should be only 1 key in all of the inputs; so if we have more than 1 then that is invalid.
+    if (uniquePublicKeys.size !== 1) {

```

Comment:
Sanity: it seems possible to have `undefined` as the only unique key.

Reply:
Yes it is possible. `undefined` is one of the expected outputs and the caller will interpret undefined === 'no public key was found' === not a valid sidetree txn.